File: articles/probability-variation-biological-inference/python/bayesian_update.py
"""
Beta-binomial Bayesian updating workflow.

Run:
    python python/bayesian_update.py
"""

from pathlib import Path

import pandas as pd

from probability_core import beta_binomial_update


ARTICLE_DIR = Path(__file__).resolve().parents[1]
PRIOR_PATH = ARTICLE_DIR / "data" / "bayesian_priors.csv"


def main() -> None:
    priors = pd.read_csv(PRIOR_PATH)

    rows = []

    for _, row in priors.iterrows():
        posterior = beta_binomial_update(
            float(row["alpha_prior"]),
            float(row["beta_prior"]),
            int(row["successes"]),
            int(row["trials"]),
        )

        rows.append(
            {
                "scenario": row["scenario"],
                "alpha_prior": row["alpha_prior"],
                "beta_prior": row["beta_prior"],
                "successes": row["successes"],
                "trials": row["trials"],
                "alpha_posterior": posterior.alpha_posterior,
                "beta_posterior": posterior.beta_posterior,
                "posterior_mean": posterior.posterior_mean,
                "posterior_sd": posterior.posterior_variance ** 0.5,
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/probability-variation-biological-inference/r/bayesian_update.R
# Beta-binomial Bayesian updating workflow in R.

prior_path <- file.path("data", "bayesian_priors.csv")
if (!file.exists(prior_path)) {
  prior_path <- file.path("..", "data", "bayesian_priors.csv")
}

priors <- read.csv(prior_path)

priors$failures <- priors$trials - priors$successes
priors$alpha_posterior <- priors$alpha_prior + priors$successes
priors$beta_posterior <- priors$beta_prior + priors$failures
priors$posterior_mean <- priors$alpha_posterior / (priors$alpha_posterior + priors$beta_posterior)
priors$posterior_sd <- sqrt(
  (priors$alpha_posterior * priors$beta_posterior) /
  ((priors$alpha_posterior + priors$beta_posterior)^2 *
   (priors$alpha_posterior + priors$beta_posterior + 1))
)

print(round(priors, 5))
